Allow includes/excludes filters for hooks

jasonwbarnett opened the pull request:
# Description

This extends the same idea used in suites to be used in lifecycle hooks, specifically the ability to include/exclude hooks based on platform name. This also goes into extracting hooks into their own class to help build a foundation for future things like https://github.com/test-kitchen/test-kitchen/issues/1641, potentially.

## This seems odd, what is the point of this and why can't you just define platform specific hooks in the platform definition?

The use case I'm trying to solve is someone who has complex platforms defined globally in a KITCHEN_GLOBAL_YAML and only defines suites in the KITCHEN_YAML file. In this case I have multiple platforms defined in the KITCHEN_GLOBAL_YAML file and I wish to define a unique pre_destroy hook per platform type (i.e. windows versus linux) in the  KITCHEN_YAML for the specific suite. Because of the way the YAML loader works with `#rmerge`, it can only merge hashes, not arrays (which totally makes sense technically). Because platforms are defined as an array it forces me to duplicate the platform content to all of my cookbooks in the mono repo.

With the feature layed out in the PR it would allow someone to do the following:

```yaml
# KITCHEN_GLOBAL_YAML
---
platforms:
  - name: redhat-7
    ...lots of content
  - name: windows-2012r2
    ...lots of content
  - name: windows-2016
    ...lots of content
```

```yaml
# KITCHEN_YAML
---
suites:
  - name: default
    run_list:
      - recipe[cookbook::default]
    pre_destroy:
      - remote: /opt/app/bin/cmd --delete
        includes:
          - redhat-7
      - remote: C:/app/bin/other-cmd --destroy
        excludes:
          - redhat-7
```

## Issues Resolved

N/A

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

clintoncwolfe commented on lib/kitchen/lifecycle_hook/base.rb:
```diff
@@ -0,0 +1,78 @@
+
+      # @return [Kitchen::StateFile]
+      def state_file
+        instance.state_file
```
When I tested this code, I get an exception thrown on this line because `state_file` is a private method of Kitchen::Instance. I reached here by having a Kitchenfile with a basic top-level lifecycle hook:
```
lifecycle:
  pre_create:
  - local: echo ....... in pre create .........
```

jasonwbarnett replied:
ahh... I haven't tested this at all at this point.

Well... using `#send` seems like a hack and probably not recommended, but what do you think could be done here? I'm not really sure given that I don't know test kitchen that well. In other words, I'm looking for some ideas here of the best way to resolve this issue.

clintoncwolfe replied:
Well, I see three options, in order of my preference:
1. The `state_file` object is actually passed in to the initialize method of LifecyleHooks. Previously, it was repeatedly passed down until it reached where it was needed. It's not graceful, but that is the least disruptive approach.
2. Use `#send` to call `state_file`
3. Alter Instance to make state_file public. I'm most reluctant to do that because it is explicitly made private.

jasonwbarnett replied:
Take a look: https://github.com/test-kitchen/test-kitchen/pull/1679/commits/013b9d5075329140510cea72b1c086eb05b17d90

clintoncwolfe reviewed (changes requested):
Thanks for your contribution! I love this feature, and I really like the approach you've taken in implementing it.  However, when I tested the code, I encountered an error which needs to be addressed. Thanks!

clintoncwolfe commented on spec/kitchen/lifecycle_hooks_spec.rb:
```diff
@@ -27,151 +27,162 @@
+    expect_local_hook_generated_and_run(:create, hook, non_standard_local_opts: { cwd: "#{kitchen_root}/test" })
     run_lifecycle_hooks
   end
 
```
As this change introduces two new options (includes and excludes) for lifecycle hooks, it would be good to add tests that cover those options.

jasonwbarnett replied:
I agree! I am going to put some time on my calendar this month to take care of this.

clintoncwolfe reviewed (commented):
I've tested the code manually and at this point it seems solid. However I think we should probably add a test to cover the new options, and ideally an update to the lifecycle hooks docs as well - https://github.com/test-kitchen/test-kitchen/blob/master/docs/content/docs/reference/lifecycle-hooks.md

jasonwbarnett commented:
@clintoncwolfe I think we're ready to rock. I added a couple tests around includes/excludes (positive and negative) and added docs to describe the new feature. Please review and let me know if you see anything else that should be done.

clintoncwolfe reviewed (approved):
Hey, sorry this sat for so long - this looks good.

jasonwbarnett commented:
I just rebased and pushed back up.

tas50 commented:
Nice work @jasonwbarnett. This is very handy
